### ISSUE TITLE
Loss is not converge

### ISSUE THREAD

**Issue by janchen0611** [rlct1/gin#1]
Hi,
I have tried your inpainting model and it is very impressive.

However, when I clone your code and try to train by myself, the loss become larger and larger after a few epochs (2~10, randomly)
I try to train from your warmup model with the same dataset(ade20k) and default parameters. But the result is still the same...

Did I do anything wrong? 
Could you give me some advice to avoid this situation?
Thanks.

**Comment by rlct1** [comment]
Hi jan, 

First of all, thanks for your appreciation. 
For your case, I have the following suggestions (may be or may not be effective):

1. keep training the model for some more time. (The training might be very unstable at the early stage)
2. extend the warmup stage. Perhaps you can try to warm up the model (the generator) with only L1 loss for few more epochs (e.g. 15 or 20 epochs) 
3. try to use gradient clipping technique (e.g. torch.nn.utils.clip_grad_norm(), refer to https://pytorch.org/docs/master/generated/torch.nn.utils.clip_grad_norm_.html) 
4. try to fix one of the two generators at the early stage of the training (e.g. fix the coarse generator and train the refinement generator for few epochs first) 

I remember that I have encountered similar problem one or two times during the early stage of the development. In my case, the loss increases suddenly at an epoch and becomes stable again at later epochs. 

I hope that these suggestions are useful for you. 
Thanks,
Ron

**Comment by junedgar** [comment]
@rlct1  
I have the same problem with janchen0611.
Can you share your train log?

**Comment by junedgar** [comment]
@janchen0611  did you get the better result with some tricks?

**Comment by rlct1** [comment]
Hi @junedgar and @janchen0611 ,

For the problem encountered, one way to deal with it is to use gradient clipping technique. 
I have updated the [train.py](https://github.com/rlct1/gin/blob/master/train.py#L104), please see line 104 for your information. Thank you very much.

Best
Ron

**Comment by junedgar** [comment]
@rlct1  the weight for style loss is default set to 80, is it reasonable?
In training time, it was alway style loss explosion.
```(epoch: 13, iters: 1104, time: 0.3618) G_GAN: 0.1199 G_COARSE_L1: 17.9970 G_OUT_L1: 413.9546 G_STYLE: 2463948800.0000 G_VGG: 74.7823 G_TV: 196.6220 D_real: 0.0265 D_fake: 1.0993```

**Comment by rlct1** [comment]
> @rlct1 the weight for style loss is default set to 80, is it reasonable?
> In training time, it was alway style loss explosion.
> `(epoch: 13, iters: 1104, time: 0.3618) G_GAN: 0.1199 G_COARSE_L1: 17.9970 G_OUT_L1: 413.9546 G_STYLE: 2463948800.0000 G_VGG: 74.7823 G_TV: 196.6220 D_real: 0.0265 D_fake: 1.0993`

The weight for style loss is defined based on [PartialConv](https://arxiv.org/pdf/1804.07723.pdf). Note that they set a default weight 120 for the style loss. As we would like to have the completed images with higher PSNR for the challenge, the dominating loss should be the L1 loss. I haven't encounter the style loss explosion before, perhaps you can try the gradient clipping technique as updated in the previous post to see whether your problem can be solved. 

Another possible solution is to take out the style loss first. Train your model without style loss to see the difference.

Thanks.
Ron

**Comment by junedgar** [comment]
@rlct1  thank you for your reply, i will follow your tricks do some experiments.
